### ISSUE TITLE
How are batches/epoch calculated?

### ISSUE THREAD

**Issue by official-elinas** [victorchall/EveryDream2trainer#65]
Coming from Koyha's trainer, it seems that a batch size of 4 at 131 epochs does 17 batches/epoch versus with Kohya, I was getting 13 batches/epoch. 

This leads to more training than I might want so I'd be interested how it works under the hood. I see you are doing a `len` calculation on line 655 in `train.py` so if you could elaborate on the parameters and how that affects batches/epoch, I would greatly appreciate it.

Also thanks for making a better trainer, this blows Kohya's out of the water with his accelerate bug.

**Comment by victorchall** [comment]
Epoch length is number of training images divided by batch size.  Ex. if you have 150 images and your batch size is 3, you'll have 50 steps per epoch (150 / 3 = 50).

If you have any uses of multiply.txt that is calculated first as it changes the "number of training images" prior to the images being batched up.

Multiaspect bucketing may also add a few steps as if you have, say, just one image at 4:1 super wide aspect, it will be multiplied automatically to fill one whole batch.   This should be pretty minor impact unless you have a vey small amount of data and lots of unique aspect ratios.  I.e. 10 images and all of them are very different aspect ratios.

If you use validation split that will remove some images from the training set.

**Comment by victorchall** [comment]
You can turn on "write_schedule" in the options and it will print out the exact images used in every step of every epoch for you into a file.  That should answer most of your questions.  Turn the option on then look at the epN_schedule.txt that it spits out in your logs. You'll see precisely what files are in each step/batch.   There is some duplication required to "fill" batches for rare aspects, you'll se "r: 2" or something next to those, that stands for "runt size", which is how I name a batch that had to be filled with duplicates.  Multiaspect requries all images in a given step/batch be identical size, and if there are "leftovers" that cannot fill a batch I duplicate then compensate on the loss calculation to avoid overfitting them.

I don't use rated dataset stuff at all, artur-ai made that and I haven't touched it.  It may be broken at this point as well and it probably wont work with multiply.txt stuff.

**Comment by nicholas-petropoulos** [comment]
Thanks for the clarification. I will give that a try.

**Comment by official-elinas** [comment]
Thanks for the clarification, I will give that a try.
